src/lib.rs

```diff
@@ -128,25 +128,33 @@ pub fn memrchr3_iter(
 /// ```
 #[inline]
 pub fn memchr(needle: u8, haystack: &[u8]) -> Option<usize> {
-    #[cfg(all(target_arch = "x86_64", memchr_runtime_simd))]
+    #[cfg(miri)]
+    #[inline(always)]
+    fn imp(n1: u8, haystack: &[u8]) -> Option<usize> {
+        naive::memchr(n1, haystack)
+    }
+
+    #[cfg(all(target_arch = "x86_64", memchr_runtime_simd, not(miri)))]
     #[inline(always)]
     fn imp(n1: u8, haystack: &[u8]) -> Option<usize> {
         x86::memchr(n1, haystack)
     }
 
     #[cfg(all(
         memchr_libc,
-        not(all(target_arch = "x86_64", memchr_runtime_simd))
+        not(all(miri, target_arch = "x86_64", memchr_runtime_simd))
     ))]
     #[inline(always)]
     fn imp(n1: u8, haystack: &[u8]) -> Option<usize> {
         c::memchr(n1, haystack)
     }
 
-    #[cfg(all(
-        not(memchr_libc),
-        not(all(target_arch = "x86_64", memchr_runtime_simd))
-    ))]
+    #[cfg(not(all(
+        miri,
+        target_arch = "x86_64",
+        memchr_runtime_simd,
+        memchr_libc
+    )))]
```

Comment:
I think this is your issue. I think you want:

```
#[cfg(all(
    not(memchr_libc),
    not(all(target_arch = "x86_64", memchr_runtime_simd)),
    not(miri),
))]
```

Reply:
Yep, `!A && !(B && C)` is `(!A && !B) || (!A && !C)` :man_facepalming: